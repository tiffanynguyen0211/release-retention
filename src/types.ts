export type Deployment = {
    Id: string;
    ReleaseId: string;
    EnvironmentId: string;
    DeployedAt: string;
};

export type Environment = {
    Id: string;
    Name: string;
};

export type Project = { Id: string; Name: string };

export type Release = {
    Id: string;
    ProjectId: string;
    Version: string | null;
    Created: string;
};

export type DisplayResult = {
    environment: Environment;
    deployment?: Deployment;
    release?: Release;
};
