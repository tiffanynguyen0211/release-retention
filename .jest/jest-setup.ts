import { configure, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

configure({ adapter: new Adapter() });

// set up axe core render base on https://github.com/dequelabs/axe-core/blob/develop/doc/examples/jest_react/test-helpers.js
// default rules on run here https://github.com/dequelabs/axe-core/blob/develop/doc/rule-descriptions.md
let wrapper: HTMLElement;
declare var jsdom: any;

export function mountToDoc(reactElm: JSX.Element) {
    if (!document) {
        // Set up a basic DOM
        global.document = jsdom('<!doctype html><html><body></body></html>');
    }
    if (!wrapper) {
        wrapper = document.createElement('main');
        document.body.appendChild(wrapper);
    }

    const container = mount(reactElm);
    wrapper.innerHTML = '';
    wrapper.appendChild(container.getDOMNode());
    return container;
}
