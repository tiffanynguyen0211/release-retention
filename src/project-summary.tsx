import * as React from 'react';
import { DisplayResult, Project } from './types';

/**
 *  Project summary dummy component that display release rentention result
 * and which deployment linked to the release to which environment
 * @example <caption>if there is/are release(s) to display as table row</caption>
 * |  Releases  | env      | deployment   |
 * |------------|----------|--------------|
 * | release-1  | env-1    | deploy-1     |
 *
 * @example <caption>if there is no release to keep display table without data rows </caption>
 * |  Releases  | env      | deployment   |
 * |------------|----------|--------------|
 *
 */
export const ProjectSummary: React.FC<{
    project: Project;
    latestDeployments: DisplayResult[] | null;
}> = ({ project, latestDeployments }) => {
    return (
        <div>
            <h1>{project.Name}</h1>
            {latestDeployments && (
                <table>
                    <thead>
                        <tr>
                            <th>Release</th>
                            <th>Environment</th>
                            <th>Deployment</th>
                        </tr>
                    </thead>
                    <tbody>
                        {latestDeployments.map(
                            ({ deployment, release, environment }) =>
                                deployment &&
                                release && (
                                    <tr key={deployment.Id}>
                                        <td>{`${release.Id} (Version:${release.Version}, Created:${release.Created})`}</td>
                                        <td>{environment.Id}</td>
                                        <td>{`${deployment.Id} (DeployedAt:${deployment.DeployedAt})`}</td>
                                    </tr>
                                ),
                        )}
                    </tbody>
                </table>
            )}
        </div>
    );
};
