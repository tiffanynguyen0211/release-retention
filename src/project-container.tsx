import * as React from 'react';
import { useEffect, useState } from 'react';
import { ProjectSummary } from './project-summary';
import {
    Deployment,
    DisplayResult,
    Environment,
    Project,
    Release,
} from './types';

/**
 *  Project Container that handles all filtering logics before passing data into actualy release retention summary
 *  assuming all data received at the beginning and only related deployments and releases are passed to release retention summary
 */
export const ProjectContainer: React.FC<{
    project: Project;
    releases?: Release[];
    environments: Environment[];
    deployments?: Deployment[];
}> = ({ project, releases = [], environments, deployments = [] }) => {
    const [latestDeployments, setLatestDeployments] = useState<
        DisplayResult[] | null
    >(null);

    // get all releases belongs to a project
    const getRelatedReleases = ({
        releases,
        projectId,
    }: {
        releases: Release[];
        projectId: string;
    }) => {
        if (releases.length === 0) return [];
        return releases.filter(release => release.ProjectId === projectId);
    };

    // get all deploys belongs to a project
    const getRelatedDeploys = ({
        relatedReleases,
        deployments,
    }: {
        relatedReleases: Release[];
        deployments: Deployment[];
    }) => {
        const relatedReleaseIds = relatedReleases.map(
            relatedRelease => relatedRelease.Id,
        );
        return deployments.filter(
            deployment => relatedReleaseIds.indexOf(deployment.ReleaseId) >= 0,
        );
    };

    // get release by Id
    const getReleaseById = (id: string, releases: Release[]) => {
        return releases.find(release => release.Id === id);
    };

    // get lastest deploy in one environment
    const getLatestDeploymentInEnv: (
        environment: Environment,
        deployments: Deployment[],
    ) => Deployment | undefined = (environment, deployments) => {
        // filter only deployments belong to the environment
        const results = deployments.filter(
            deployment => deployment.EnvironmentId === environment.Id,
        );

        //if no deploys found for the env, return undefined
        if (results.length === 0) return undefined;

        // else order by DeployedAt DESC
        results.sort(function(x, y) {
            // if found multiple deploys deployed at the sametime
            // rare case but not impossible
            if (Date.parse(y.DeployedAt) === Date.parse(x.DeployedAt)) {
                // comparing by Id - assuming every deploy Id is creared incrementally
                // return ordering by Id DESC
                return (
                    parseInt(y.Id.split('-')[1]) - parseInt(x.Id.split('-')[1])
                );
            }
            return Date.parse(y.DeployedAt) - Date.parse(x.DeployedAt);
        });

        // only latest deployment is selected
        return results[0];
    };

    // get lastest deploy in each environment that belong to the release of the project
    const getLatestDeploymentForEachEnv = ({
        deployments,
        envs,
        releases,
    }: {
        deployments: Deployment[];
        envs: Environment[];
        releases: Release[];
    }) => {
        const results: DisplayResult[] = [];

        envs.forEach(env => {
            // get latest deploy in the environment from project's deployments (can be undefined)
            const latestDeploy = getLatestDeploymentInEnv(env, deployments);

            // if there is a latest deploy in the env, find the release
            // ( can be undefined if the release is no longer existed in the project releases for some reason)
            const releaseByDeploy = latestDeploy
                ? getReleaseById(latestDeploy.ReleaseId, releases)
                : undefined;

            // if the release existed then assign the deploy to display result
            const result = releaseByDeploy ? latestDeploy : undefined;

            results.push({
                environment: env,
                deployment: result,
                release: releaseByDeploy,
            });
        });

        return results;
    };

    useEffect(() => {
        // get all releases belong to the project (can be empty array)
        const filteredReleases = getRelatedReleases({
            releases,
            projectId: project.Id,
        });

        // get all deploys belong to the project (can be empty array)
        const filteredDeploys = getRelatedDeploys({
            relatedReleases: filteredReleases,
            deployments,
        });

        // if there is no deploy to the project found the no point going to the next step
        // summary of release retion will display empty table
        if (filteredDeploys.length === 0) {
            setLatestDeployments(null);
            return;
        }

        // filter out the latest deploy for each environment of the project
        const results = getLatestDeploymentForEachEnv({
            deployments: filteredDeploys,
            envs: environments,
            releases: filteredReleases,
        });

        // set the released retention result
        setLatestDeployments(results);
    }, [releases, deployments, environments]);

    useEffect(() => {
        // this result log should only run in production build
        if (
            latestDeployments &&
            latestDeployments.length > 0 &&
            process.env.NODE_ENV === 'production'
        ) {
            latestDeployments.map(({ environment, release }) => {
                if (release) {
                    // @TODO intergrate with api to log result
                    console.log(
                        `${project.Id}: ${release.Id} kept because it was the most recently deployed to ${environment.Id}`,
                    );
                }
            });
        }
    }, [latestDeployments]);

    return (
        <ProjectSummary
            project={project}
            latestDeployments={latestDeployments}
        />
    );
};
