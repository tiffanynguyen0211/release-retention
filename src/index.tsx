import * as React from 'react';
import { ProjectContainer } from './project-container';
// import { ProjectSummary } from './project-summary';
import { Deployment, Environment, Project, Release } from './types';

/**
 *  React Component that display all projects with release retention info in each project
 *  assuming all data received at the beginning
 */
export const ReleaseRetension: React.FC<{
    projects?: Project[];
    releases?: Release[];
    environments?: Environment[];
    deployments?: Deployment[];
}> = ({ projects, releases = [], environments, deployments = [] }) => {
    // render empty react fragment if there is no project or environments data
    // assuming this feature is not handling display all releases without environment ready
    if (!projects || !environments || environments.length === 0) return <></>;

    return (
        <div>
            {projects.map(project => (
                <ProjectContainer
                    key={project.Id}
                    project={project}
                    releases={releases}
                    deployments={deployments}
                    environments={environments}
                />
            ))}
        </div>
    );
};
