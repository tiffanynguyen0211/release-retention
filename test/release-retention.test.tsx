import * as React from 'react';
import axe from 'axe-core';
import { mount, ReactWrapper, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ReactElement } from 'react';

import releaseData from '../staticData/Releases.json';
import projectData from '../staticData/Projects.json';
import deploymentData from '../staticData/Deployments.json';
import environmentData from '../staticData/Environments.json';

import { ReleaseRetension } from '../src';
import { mountToDoc } from '../.jest/jest-setup';
import { Deployment, Environment, Project, Release } from '../src/types';

const setup = (
    render: (x: ReactElement) => ReactWrapper | ShallowWrapper,
    props: JSX.IntrinsicAttributes,
) => {
    const defaultProps: {
        projects?: Project[];
        releases?: Release[];
        environments?: Environment[];
        deployments?: Deployment[];
    } = {
        projects: projectData,
        releases: releaseData,
        environments: environmentData,
        deployments: deploymentData,
    };

    const component = render(<ReleaseRetension {...defaultProps} {...props} />);

    return {
        actual: component,
    };
};

describe('Release Retention', () => {
    it('should match snapshot on first render', () => {
        const { actual } = setup(mount, {});
        actual.update();
        expect(toJson(actual)).toMatchSnapshot();
    });

    it('has no accessibility violations', () => {
        const { actual } = setup(mountToDoc, {});
        const componentNode = actual.getDOMNode();

        axe.run(componentNode, (err, { violations }) => {
            expect(err).toBe(null);
            expect(violations).toHaveLength(0);
        });
    });
});
