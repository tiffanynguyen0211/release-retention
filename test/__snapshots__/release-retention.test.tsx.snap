// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Release Retention should match snapshot on first render 1`] = `
<ReleaseRetension
  deployments={
    Array [
      Object {
        "DeployedAt": "2000-01-02T10:00:00",
        "EnvironmentId": "Environment-1",
        "Id": "Deployment-2",
        "ReleaseId": "Release-2",
      },
      Object {
        "DeployedAt": "2000-01-02T11:00:00",
        "EnvironmentId": "Environment-2",
        "Id": "Deployment-3",
        "ReleaseId": "Release-1",
      },
      Object {
        "DeployedAt": "2000-01-02T12:00:00",
        "EnvironmentId": "Environment-3",
        "Id": "Deployment-4",
        "ReleaseId": "Release-2",
      },
      Object {
        "DeployedAt": "2000-01-01T11:00:00",
        "EnvironmentId": "Environment-1",
        "Id": "Deployment-5",
        "ReleaseId": "Release-5",
      },
      Object {
        "DeployedAt": "2000-01-02T10:00:00",
        "EnvironmentId": "Environment-1",
        "Id": "Deployment-6",
        "ReleaseId": "Release-6",
      },
      Object {
        "DeployedAt": "2000-01-02T11:00:00",
        "EnvironmentId": "Environment-2",
        "Id": "Deployment-7",
        "ReleaseId": "Release-6",
      },
      Object {
        "DeployedAt": "2000-01-02T13:00:00",
        "EnvironmentId": "Environment-1",
        "Id": "Deployment-8",
        "ReleaseId": "Release-7",
      },
      Object {
        "DeployedAt": "2000-01-02T14:00:00",
        "EnvironmentId": "Environment-1",
        "Id": "Deployment-9",
        "ReleaseId": "Release-6",
      },
      Object {
        "DeployedAt": "2000-01-01T10:00:00",
        "EnvironmentId": "Environment-1",
        "Id": "Deployment-10",
        "ReleaseId": "Release-8",
      },
    ]
  }
  environments={
    Array [
      Object {
        "Id": "Environment-1",
        "Name": "Staging",
      },
      Object {
        "Id": "Environment-2",
        "Name": "Production",
      },
    ]
  }
  projects={
    Array [
      Object {
        "Id": "Project-1",
        "Name": "Random Quotes",
      },
      Object {
        "Id": "Project-2",
        "Name": "Pet Shop",
      },
      Object {
        "Id": "Project-3",
        "Name": "Is this project is missed?",
      },
    ]
  }
  releases={
    Array [
      Object {
        "Created": "2000-01-01T09:00:00",
        "Id": "Release-1",
        "ProjectId": "Project-1",
        "Version": "1.0.0",
      },
      Object {
        "Created": "2000-01-02T09:00:00",
        "Id": "Release-2",
        "ProjectId": "Project-1",
        "Version": "1.0.1",
      },
      Object {
        "Created": "2000-01-02T13:00:00",
        "Id": "Release-3",
        "ProjectId": "Project-1",
        "Version": null,
      },
      Object {
        "Created": "2000-01-01T09:00:00",
        "Id": "Release-4",
        "ProjectId": "Project-2",
        "Version": "1.0.0",
      },
      Object {
        "Created": "2000-01-01T10:00:00",
        "Id": "Release-5",
        "ProjectId": "Project-2",
        "Version": "1.0.1-ci1",
      },
      Object {
        "Created": "2000-01-02T09:00:00",
        "Id": "Release-6",
        "ProjectId": "Project-2",
        "Version": "1.0.2",
      },
      Object {
        "Created": "2000-01-02T12:00:00",
        "Id": "Release-7",
        "ProjectId": "Project-2",
        "Version": "1.0.3",
      },
      Object {
        "Created": "2000-01-01T09:00:00",
        "Id": "Release-8",
        "ProjectId": "Project-3",
        "Version": "2.0.0",
      },
    ]
  }
>
  <div>
    <ProjectContainer
      deployments={
        Array [
          Object {
            "DeployedAt": "2000-01-02T10:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-2",
            "ReleaseId": "Release-2",
          },
          Object {
            "DeployedAt": "2000-01-02T11:00:00",
            "EnvironmentId": "Environment-2",
            "Id": "Deployment-3",
            "ReleaseId": "Release-1",
          },
          Object {
            "DeployedAt": "2000-01-02T12:00:00",
            "EnvironmentId": "Environment-3",
            "Id": "Deployment-4",
            "ReleaseId": "Release-2",
          },
          Object {
            "DeployedAt": "2000-01-01T11:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-5",
            "ReleaseId": "Release-5",
          },
          Object {
            "DeployedAt": "2000-01-02T10:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-6",
            "ReleaseId": "Release-6",
          },
          Object {
            "DeployedAt": "2000-01-02T11:00:00",
            "EnvironmentId": "Environment-2",
            "Id": "Deployment-7",
            "ReleaseId": "Release-6",
          },
          Object {
            "DeployedAt": "2000-01-02T13:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-8",
            "ReleaseId": "Release-7",
          },
          Object {
            "DeployedAt": "2000-01-02T14:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-9",
            "ReleaseId": "Release-6",
          },
          Object {
            "DeployedAt": "2000-01-01T10:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-10",
            "ReleaseId": "Release-8",
          },
        ]
      }
      environments={
        Array [
          Object {
            "Id": "Environment-1",
            "Name": "Staging",
          },
          Object {
            "Id": "Environment-2",
            "Name": "Production",
          },
        ]
      }
      key="Project-1"
      project={
        Object {
          "Id": "Project-1",
          "Name": "Random Quotes",
        }
      }
      releases={
        Array [
          Object {
            "Created": "2000-01-01T09:00:00",
            "Id": "Release-1",
            "ProjectId": "Project-1",
            "Version": "1.0.0",
          },
          Object {
            "Created": "2000-01-02T09:00:00",
            "Id": "Release-2",
            "ProjectId": "Project-1",
            "Version": "1.0.1",
          },
          Object {
            "Created": "2000-01-02T13:00:00",
            "Id": "Release-3",
            "ProjectId": "Project-1",
            "Version": null,
          },
          Object {
            "Created": "2000-01-01T09:00:00",
            "Id": "Release-4",
            "ProjectId": "Project-2",
            "Version": "1.0.0",
          },
          Object {
            "Created": "2000-01-01T10:00:00",
            "Id": "Release-5",
            "ProjectId": "Project-2",
            "Version": "1.0.1-ci1",
          },
          Object {
            "Created": "2000-01-02T09:00:00",
            "Id": "Release-6",
            "ProjectId": "Project-2",
            "Version": "1.0.2",
          },
          Object {
            "Created": "2000-01-02T12:00:00",
            "Id": "Release-7",
            "ProjectId": "Project-2",
            "Version": "1.0.3",
          },
          Object {
            "Created": "2000-01-01T09:00:00",
            "Id": "Release-8",
            "ProjectId": "Project-3",
            "Version": "2.0.0",
          },
        ]
      }
    >
      <ProjectSummary
        latestDeployments={
          Array [
            Object {
              "deployment": Object {
                "DeployedAt": "2000-01-02T10:00:00",
                "EnvironmentId": "Environment-1",
                "Id": "Deployment-2",
                "ReleaseId": "Release-2",
              },
              "environment": Object {
                "Id": "Environment-1",
                "Name": "Staging",
              },
              "release": Object {
                "Created": "2000-01-02T09:00:00",
                "Id": "Release-2",
                "ProjectId": "Project-1",
                "Version": "1.0.1",
              },
            },
            Object {
              "deployment": Object {
                "DeployedAt": "2000-01-02T11:00:00",
                "EnvironmentId": "Environment-2",
                "Id": "Deployment-3",
                "ReleaseId": "Release-1",
              },
              "environment": Object {
                "Id": "Environment-2",
                "Name": "Production",
              },
              "release": Object {
                "Created": "2000-01-01T09:00:00",
                "Id": "Release-1",
                "ProjectId": "Project-1",
                "Version": "1.0.0",
              },
            },
          ]
        }
        project={
          Object {
            "Id": "Project-1",
            "Name": "Random Quotes",
          }
        }
      >
        <div>
          <h1>
            Random Quotes
          </h1>
          <table>
            <thead>
              <tr>
                <th>
                  Release
                </th>
                <th>
                  Environment
                </th>
                <th>
                  Deployment
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                key="Deployment-2"
              >
                <td>
                  Release-2 (Version:1.0.1, Created:2000-01-02T09:00:00)
                </td>
                <td>
                  Environment-1
                </td>
                <td>
                  Deployment-2 (DeployedAt:2000-01-02T10:00:00)
                </td>
              </tr>
              <tr
                key="Deployment-3"
              >
                <td>
                  Release-1 (Version:1.0.0, Created:2000-01-01T09:00:00)
                </td>
                <td>
                  Environment-2
                </td>
                <td>
                  Deployment-3 (DeployedAt:2000-01-02T11:00:00)
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </ProjectSummary>
    </ProjectContainer>
    <ProjectContainer
      deployments={
        Array [
          Object {
            "DeployedAt": "2000-01-02T10:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-2",
            "ReleaseId": "Release-2",
          },
          Object {
            "DeployedAt": "2000-01-02T11:00:00",
            "EnvironmentId": "Environment-2",
            "Id": "Deployment-3",
            "ReleaseId": "Release-1",
          },
          Object {
            "DeployedAt": "2000-01-02T12:00:00",
            "EnvironmentId": "Environment-3",
            "Id": "Deployment-4",
            "ReleaseId": "Release-2",
          },
          Object {
            "DeployedAt": "2000-01-01T11:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-5",
            "ReleaseId": "Release-5",
          },
          Object {
            "DeployedAt": "2000-01-02T10:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-6",
            "ReleaseId": "Release-6",
          },
          Object {
            "DeployedAt": "2000-01-02T11:00:00",
            "EnvironmentId": "Environment-2",
            "Id": "Deployment-7",
            "ReleaseId": "Release-6",
          },
          Object {
            "DeployedAt": "2000-01-02T13:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-8",
            "ReleaseId": "Release-7",
          },
          Object {
            "DeployedAt": "2000-01-02T14:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-9",
            "ReleaseId": "Release-6",
          },
          Object {
            "DeployedAt": "2000-01-01T10:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-10",
            "ReleaseId": "Release-8",
          },
        ]
      }
      environments={
        Array [
          Object {
            "Id": "Environment-1",
            "Name": "Staging",
          },
          Object {
            "Id": "Environment-2",
            "Name": "Production",
          },
        ]
      }
      key="Project-2"
      project={
        Object {
          "Id": "Project-2",
          "Name": "Pet Shop",
        }
      }
      releases={
        Array [
          Object {
            "Created": "2000-01-01T09:00:00",
            "Id": "Release-1",
            "ProjectId": "Project-1",
            "Version": "1.0.0",
          },
          Object {
            "Created": "2000-01-02T09:00:00",
            "Id": "Release-2",
            "ProjectId": "Project-1",
            "Version": "1.0.1",
          },
          Object {
            "Created": "2000-01-02T13:00:00",
            "Id": "Release-3",
            "ProjectId": "Project-1",
            "Version": null,
          },
          Object {
            "Created": "2000-01-01T09:00:00",
            "Id": "Release-4",
            "ProjectId": "Project-2",
            "Version": "1.0.0",
          },
          Object {
            "Created": "2000-01-01T10:00:00",
            "Id": "Release-5",
            "ProjectId": "Project-2",
            "Version": "1.0.1-ci1",
          },
          Object {
            "Created": "2000-01-02T09:00:00",
            "Id": "Release-6",
            "ProjectId": "Project-2",
            "Version": "1.0.2",
          },
          Object {
            "Created": "2000-01-02T12:00:00",
            "Id": "Release-7",
            "ProjectId": "Project-2",
            "Version": "1.0.3",
          },
          Object {
            "Created": "2000-01-01T09:00:00",
            "Id": "Release-8",
            "ProjectId": "Project-3",
            "Version": "2.0.0",
          },
        ]
      }
    >
      <ProjectSummary
        latestDeployments={
          Array [
            Object {
              "deployment": Object {
                "DeployedAt": "2000-01-02T14:00:00",
                "EnvironmentId": "Environment-1",
                "Id": "Deployment-9",
                "ReleaseId": "Release-6",
              },
              "environment": Object {
                "Id": "Environment-1",
                "Name": "Staging",
              },
              "release": Object {
                "Created": "2000-01-02T09:00:00",
                "Id": "Release-6",
                "ProjectId": "Project-2",
                "Version": "1.0.2",
              },
            },
            Object {
              "deployment": Object {
                "DeployedAt": "2000-01-02T11:00:00",
                "EnvironmentId": "Environment-2",
                "Id": "Deployment-7",
                "ReleaseId": "Release-6",
              },
              "environment": Object {
                "Id": "Environment-2",
                "Name": "Production",
              },
              "release": Object {
                "Created": "2000-01-02T09:00:00",
                "Id": "Release-6",
                "ProjectId": "Project-2",
                "Version": "1.0.2",
              },
            },
          ]
        }
        project={
          Object {
            "Id": "Project-2",
            "Name": "Pet Shop",
          }
        }
      >
        <div>
          <h1>
            Pet Shop
          </h1>
          <table>
            <thead>
              <tr>
                <th>
                  Release
                </th>
                <th>
                  Environment
                </th>
                <th>
                  Deployment
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                key="Deployment-9"
              >
                <td>
                  Release-6 (Version:1.0.2, Created:2000-01-02T09:00:00)
                </td>
                <td>
                  Environment-1
                </td>
                <td>
                  Deployment-9 (DeployedAt:2000-01-02T14:00:00)
                </td>
              </tr>
              <tr
                key="Deployment-7"
              >
                <td>
                  Release-6 (Version:1.0.2, Created:2000-01-02T09:00:00)
                </td>
                <td>
                  Environment-2
                </td>
                <td>
                  Deployment-7 (DeployedAt:2000-01-02T11:00:00)
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </ProjectSummary>
    </ProjectContainer>
    <ProjectContainer
      deployments={
        Array [
          Object {
            "DeployedAt": "2000-01-02T10:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-2",
            "ReleaseId": "Release-2",
          },
          Object {
            "DeployedAt": "2000-01-02T11:00:00",
            "EnvironmentId": "Environment-2",
            "Id": "Deployment-3",
            "ReleaseId": "Release-1",
          },
          Object {
            "DeployedAt": "2000-01-02T12:00:00",
            "EnvironmentId": "Environment-3",
            "Id": "Deployment-4",
            "ReleaseId": "Release-2",
          },
          Object {
            "DeployedAt": "2000-01-01T11:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-5",
            "ReleaseId": "Release-5",
          },
          Object {
            "DeployedAt": "2000-01-02T10:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-6",
            "ReleaseId": "Release-6",
          },
          Object {
            "DeployedAt": "2000-01-02T11:00:00",
            "EnvironmentId": "Environment-2",
            "Id": "Deployment-7",
            "ReleaseId": "Release-6",
          },
          Object {
            "DeployedAt": "2000-01-02T13:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-8",
            "ReleaseId": "Release-7",
          },
          Object {
            "DeployedAt": "2000-01-02T14:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-9",
            "ReleaseId": "Release-6",
          },
          Object {
            "DeployedAt": "2000-01-01T10:00:00",
            "EnvironmentId": "Environment-1",
            "Id": "Deployment-10",
            "ReleaseId": "Release-8",
          },
        ]
      }
      environments={
        Array [
          Object {
            "Id": "Environment-1",
            "Name": "Staging",
          },
          Object {
            "Id": "Environment-2",
            "Name": "Production",
          },
        ]
      }
      key="Project-3"
      project={
        Object {
          "Id": "Project-3",
          "Name": "Is this project is missed?",
        }
      }
      releases={
        Array [
          Object {
            "Created": "2000-01-01T09:00:00",
            "Id": "Release-1",
            "ProjectId": "Project-1",
            "Version": "1.0.0",
          },
          Object {
            "Created": "2000-01-02T09:00:00",
            "Id": "Release-2",
            "ProjectId": "Project-1",
            "Version": "1.0.1",
          },
          Object {
            "Created": "2000-01-02T13:00:00",
            "Id": "Release-3",
            "ProjectId": "Project-1",
            "Version": null,
          },
          Object {
            "Created": "2000-01-01T09:00:00",
            "Id": "Release-4",
            "ProjectId": "Project-2",
            "Version": "1.0.0",
          },
          Object {
            "Created": "2000-01-01T10:00:00",
            "Id": "Release-5",
            "ProjectId": "Project-2",
            "Version": "1.0.1-ci1",
          },
          Object {
            "Created": "2000-01-02T09:00:00",
            "Id": "Release-6",
            "ProjectId": "Project-2",
            "Version": "1.0.2",
          },
          Object {
            "Created": "2000-01-02T12:00:00",
            "Id": "Release-7",
            "ProjectId": "Project-2",
            "Version": "1.0.3",
          },
          Object {
            "Created": "2000-01-01T09:00:00",
            "Id": "Release-8",
            "ProjectId": "Project-3",
            "Version": "2.0.0",
          },
        ]
      }
    >
      <ProjectSummary
        latestDeployments={
          Array [
            Object {
              "deployment": Object {
                "DeployedAt": "2000-01-01T10:00:00",
                "EnvironmentId": "Environment-1",
                "Id": "Deployment-10",
                "ReleaseId": "Release-8",
              },
              "environment": Object {
                "Id": "Environment-1",
                "Name": "Staging",
              },
              "release": Object {
                "Created": "2000-01-01T09:00:00",
                "Id": "Release-8",
                "ProjectId": "Project-3",
                "Version": "2.0.0",
              },
            },
            Object {
              "deployment": undefined,
              "environment": Object {
                "Id": "Environment-2",
                "Name": "Production",
              },
              "release": undefined,
            },
          ]
        }
        project={
          Object {
            "Id": "Project-3",
            "Name": "Is this project is missed?",
          }
        }
      >
        <div>
          <h1>
            Is this project is missed?
          </h1>
          <table>
            <thead>
              <tr>
                <th>
                  Release
                </th>
                <th>
                  Environment
                </th>
                <th>
                  Deployment
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                key="Deployment-10"
              >
                <td>
                  Release-8 (Version:2.0.0, Created:2000-01-01T09:00:00)
                </td>
                <td>
                  Environment-1
                </td>
                <td>
                  Deployment-10 (DeployedAt:2000-01-01T10:00:00)
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </ProjectSummary>
    </ProjectContainer>
  </div>
</ReleaseRetension>
`;
